### ISSUE TITLE
FUNCTIONS_V2_COMPATIBILITY_MODE

### ISSUE THREAD

**Issue by vitalybibikov** [Azure/azure-functions-host#6399]
There is no documentation on how exactly FUNCTIONS_V2_COMPATIBILITY_MODE = true works at the moment.
Also, there is no any info on when and in what circumstances it can be deleted for v3.
As currently, without it apps simply does not work in Azure.

**Comment by mhoeger** [comment]
Hi @vitalybibikov - This setting is in place as an internal setting to be able to migrate functions v2 apps off of .net core 2.

Whenever you have FUNCTIONS_V2_COMPATIBILITY_MODE enabled, the breaking changes in 3.x are reverted: https://docs.microsoft.com/en-us/azure/azure-functions/functions-versions#breaking-changes-between-2x-and-3x

I might be wrong, but this isn't an official setting we want to support and document (@fabiocav correct me if I'm wrong!)

**Comment by fabiocav** [comment]
@mhoeger's comment is correct. This was exposed as a way to unblock migration to 3.0, but should be looked as a temporary solution while you mitigate the underlying issue.

**Comment by fabiocav** [comment]
If you can provide more information about the errors you see without the flag, we can provide some guidance on how to address them.

**Comment by vitalybibikov** [comment]
I've found out that it fails only in Azure.
And it fails for all Commands(POST,PuT,etc) functions due to the fact that I'm deserializing model out of request myself (currently out-of the box deserialization does not handle Enums/Lists/etc) properly.

Error in AppInsights:

```
Unhandled exception: System.Text.Json.JsonException: The input does not contain any JSON tokens. Expected the input to start with a valid JSON token, when isFinalBlock is true. Path: $ | LineNumber: 0 | BytePositionInLine: 0.
 ---> System.Text.Json.JsonReaderException: The input does not contain any JSON tokens. Expected the input to start with a valid JSON token, when isFinalBlock is true. LineNumber: 0 | BytePositionInLine: 0.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack)
```

And it fails in **prod** on this code 

```
            var body = await request.Content.ReadAsStringAsync();
            var result = JsonSerializer.Deserialize<T>(body);
```


After adding some logging, I can see that HttpRequest Body Length is 0
However, If I change HttpRequest to HttpRequestMessage, then body  length is "40" and no longer equals 0.

So, it must be an issue with HttpRequest Body somehow? Like improper stream position or something.

Unfortunately, as far as I can see it's not possible to have HttpRequestMessage and Model simultaneously as a function parameters,
while it is possible for HttpRequest.

Does not work:
```
            [RequestBodyType(typeof(UserProfileViewModel), "UserProfileViewModel")] UserProfileViewModel model,
            HttpRequestMessage req,
```

Works fine:
```
            [RequestBodyType(typeof(UserProfileViewModel), "UserProfileViewModel")] UserProfileViewModel model,
            HttpMessage req,
```

Is this also a bug?

**Comment by fabiocav** [comment]
You can have the request and the model at the same time. All you need to to is apply the trigger attribute to the POCO parameter, while also having a parameter that takes an `HttpRequest`

**Comment by vitalybibikov** [comment]
As I mentioned above:
It does not work.

Does not work:
```
            [RequestBodyType(typeof(UserProfileViewModel), "UserProfileViewModel")] UserProfileViewModel model,
            HttpRequestMessage req,
```

Works fine:
```
            [RequestBodyType(typeof(UserProfileViewModel), "UserProfileViewModel")] UserProfileViewModel model,
            HttpMessage req,
```

And it's another issue from that one that I mentioned in the first place.

**Comment by fabiocav** [comment]
Can you please provide a repro do demonstrate your issue?

`RequestBodyType` is not supported in Azure Functions. In order to bind to the request and the model (POCO), you need to follow the approach I've described above. Does that work for you? In addition to `HttpRequest`, the request parameter may also use `HttpRequestMessage` (this is to provide backwards compatibility)
